onchange_helper/models/models.py

```diff
@@ -1,51 +1,170 @@
-        new_values = {}
-        for fieldname, value in vals.iteritems():
-            if fieldname not in record:
-                column = self._fields[fieldname]
-                if value and column.type == 'many2one':
-                    value = value[0]  # many2one are tuple (id, name)
-                new_values[fieldname] = value
-        return new_values
-
-    def play_onchanges(self, values, onchange_fields):
-        onchange_specs = self._onchange_spec()
+    _inherit = "base"
+
+    @api.multi
+    def _compute_onchange_dirty(self, original_record):
+        """
+        Return the list of dirty fields. (designed to be called by
+        play_onchanges)
+        The list of dirty fields is computed from the list marked as dirty
+        on the record. Form this list, we remove the fields for which the value
+        into the original record is the same as the one into the current record
+        :param original_record:
+        :return: changed values
+        """
+        self.ensure_one()
+        dirties = []
+        for field_name in self._get_dirty():
+            original_value = original_record[field_name]
+            new_value = self[field_name]
+            if original_value == new_value:
+                # False dirty. The field has been loaded into the
+                # onchange process -> skip
+                continue
+            dirties.append(field_name)
+        return dirties
+
+    def play_onchanges(self, values, onchange_fields=None):
+        """
+        Play the onchange methods defined on the current record and return the
+        changed values.
+        The record is not modified by the onchange
+
+        :param values: dict of input value that
+        :param onchange_fields: fields for which onchange methods will be
+        played. If not provided, the list of field is based on the values keys.
+        Order in onchange_fields is very important as onchanges methods will
+        be played in that order.
+        :return: changed values
+
+        This method reimplement the onchange method to be able to work on the
+        current recordset if provided.
+        """
+        env = self.env
+        if self:
+            self.ensure_one()
+
+        if not onchange_fields:
+            onchange_fields = values.keys()
+        names = onchange_fields
+
+        # _onchange_spec() will return onchange fields from the default view
         # we need all fields in the dict even the empty ones
         # otherwise 'onchange()' will not apply changes to them
-        all_values = values.copy()
+        field_onchange = {field_name: "1" for field_name in self._fields}
+
+        if not all(name in self._fields for name in onchange_fields):
+            return {}
+
+        # determine subfields for field.convert_to_onchange() below
+        secondary = []
+        subfields = defaultdict(set)
+        for dotname in field_onchange:
+            if "." in dotname:
```

Comment:
@mourad-ehm I'll remove all the code related to subfields since the code doesn't rely on a FormView. We'll never have a dot into a `field_onchange` key ...

Reply:
OK